### PR TITLE
Feature gating subsystems via `#[cfg]`

### PR DESCRIPTION
**Pull request by skunert**
In this PR I introduce guarding of subsystems via features.
Supported syntax:
```rust
#[orchestra(signal=Signal, event=ExternalEvent, error=SubsystemError, gen=AllMessages)]
struct MyOverseer<T, U, V, W> {
	#[cfg(feature = "feature1")]
	#[subsystem(consumes: Subsystem1Message, sends: [Subsystem2Message])]
	subsystem1: Sub1,

	#[subsystem(consumes: Subsystem2Message, sends: [Subsystem1Message])]
	subsystem2: Sub2,
}
```
Basically this allows re-use of one orchestra in scenarios where it is not needed to run all subsystems. The `#cfg` attribute macro currently supports `any`, `all`, `not` and `feature = "something"`.

## Limitations
- Feature unification. The current implementation is not very useful when you have a large workspace, and different crates depend on the defined orchestra with different features enabled.

- Performance depends on the number of unique cfg expressions defined in the crate. With the current design of the orchestra builder, I have to build distinct feature combinations for the builder. This scales with O(2^n) where n is the number of distinct feature expressions. I tested a bit and impact is starting to show at around 7 different feature expressions.

### CODE REVIEW

File: orchestra/proc-macro/src/impl_builder.rs [hunk 1]
```diff
@@ -31,26 +31,144 @@ fn recollect_without_idx<T: Clone>(x: &[T], idx: usize) -> Vec<T> {
...
 pub(crate) fn impl_builder(info: &OrchestraInfo) -> proc_macro2::TokenStream {
 	let orchestra_name = info.orchestra_name.clone();
+
 	let builder = format_ident!("{}Builder", orchestra_name);
 	let handle = format_ident!("{}Handle", orchestra_name);
 	let connector = format_ident!("{}Connector", orchestra_name);
 	let subsystem_ctx_name = format_ident!("{}SubsystemContext", orchestra_name);
+	let feature_powerset_cfgs = &info.feature_gated_subsystem_sets();
+
+	let mut ts = quote! {};
```
Comment (drahnr) [inline orchestra/proc-macro/src/impl_builder.rs:41]:
```suggestion
	let mut ts = proc_macro2::TokenStream::new();
```

File: orchestra/proc-macro/src/impl_builder.rs [hunk 2]
```diff
@@ -31,26 +31,144 @@ fn recollect_without_idx<T: Clone>(x: &[T], idx: usize) -> Vec<T> {
...
+			}
+	});
+	ts.extend(impl_task_kind(info));
+	ts
+}
 
-	let consumes = &info.consumes_without_wip();
-	let channel_name = &info.channel_names_without_wip(None);
-	let channel_name_unbounded = &info.channel_names_without_wip("_unbounded");
+pub(crate) fn impl_feature_guarded_items(
```
Comment (drahnr) [inline orchestra/proc-macro/src/impl_builder.rs:146]:
```suggestion
pub(crate) fn impl_feature_gated_items(
```

File: orchestra/proc-macro/src/parse/parse_orchestra_struct.rs [hunk 3]
```diff
@@ -594,6 +648,34 @@ impl OrchestraGuts {
...
+
+			let feature_gates = if let Some((attr_tokens, span)) = cfg_attr.next() {
+				// multiple `#[cfg(..)]` found
+				if let Some((_attr_tokens2, span2)) = cfg_attr.next() {
+					return Err({
+						let mut err = Error::new(span, "The first cfg annotation is at");
+						err.combine(Error::new(span2, "but another here for the same field."));
+						err
+					})
+				}
```
Comment (drahnr) [inline orchestra/proc-macro/src/parse/parse_orchestra_struct.rs:670]: Shouldn't this just recombine as an `and`?

File: orchestra/proc-macro/src/parse/parse_cfg.rs [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+impl Parse for CfgExpressionRoot {
+	fn parse(input: ParseStream) -> Result<Self> {
+		let content;
+		let _paren_token = parenthesized!(content in input);
+
+		Ok(Self { predicate: content.parse()? })
+	}
+}
+
+/// Single cfg predicate for parsing.
```
Comment (drahnr) [inline orchestra/proc-macro/src/parse/parse_cfg.rs:46]: Giving a full example of an instance that'd parse successfully would be nice

File: orchestra/proc-macro/src/parse/parse_cfg.rs [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+			any(
+				not(feature = "no"),
+				any(feature = "any2", feature = "any1"),
+				all(feature = "f2", feature = "f1", feature = "f3")
+			)
+		};
+		assert_eq!("any (not (feature = \"no\") , any (feature = \"any2\" , feature = \"any1\") , all (feature = \"f2\" , feature = \"f1\" , feature = \"f3\"))"
+				   , to_parse.to_token_stream().to_string());
+	}
+}
```
Comment (drahnr) [inline orchestra/proc-macro/src/parse/parse_cfg.rs:188]: Very very nice.

**Review by drahnr — APPROVED** [review]
A few small nits, generally looks very good!

Q: do we want to allow arbitrary `#[]` macros eventually? I.e. copying `derive` over to the generated `struct` only and keeping any other ones on all items. But that's for another time I think.

File: orchestra/proc-macro/src/parse/parse_orchestra_struct.rs (hunk 3 above)
Comment (skunert) [reply orchestra/proc-macro/src/parse/parse_orchestra_struct.rs]: Absolutely correct! Done!

**Comment by skunert** [comment]
> Q: do we want to allow arbitrary #[] macros eventually? I.e. copying derive over to the generated struct only and keeping any other ones on all items. But that's for another time I think.

Not sure I understand, how could this be used?

**Review by vstakhov — APPROVED** [review]

**Pull request merged**